### ISSUE TITLE
sklearn vs xgboost inconsistency in prediction probabilities

### ISSUE THREAD

**Issue by humzaiqbal** [dmlc/xgboost#4607]
### System setup
OS: Mac OS X 10.13.6
xgboost version: 0.82
scikit-learn version: 0.20.2
numpy version: 1.16.1

### Problem
When I try to train a model using some sample data I notice that while both models are able to learn the data, the xgboost model is not able to do so as confidently judging from the prediction probabilities. 

### Code
```
import xgboost as xgb
from sklearn.ensemble import GradientBoostingClassifier
import numpy as np 

XGB_PARAM = {
    "learning_rate": 0.1,
    "n_estimators": 100,
    "objective": "binary:logistic",
    "verbosity": 1,
    "booster": "gbtree",
    "max_depth": 3,
    "lambda": 0,
}

features = np.load('features.npy')
labels = np.load('labels.npy')


def run_classifier(mode='xgb'):
	if mode == 'xgb':
		model = xgb.XGBClassifier(**XGB_PARAM)
	else:
		model = GradientBoostingClassifier()

	model.fit(features, labels)
	print('{} probabilities: '.format(mode), model.predict_proba(features))


run_classifier('xgb')
run_classifier('sklearn')
``` 
### Results 
When I run the above code I get the following result 
```
xgb probabilities:  [[0.33988953 0.6601105 ]
 [0.80120075 0.19879927]
 [0.84622085 0.15377913]
 [0.34065586 0.65934414]
 [0.8017439  0.19825609]
 [0.8017439  0.19825609]
 [0.80120075 0.19879927]
 [0.80120075 0.19879927]
 [0.33988953 0.6601105 ]
 [0.8453405  0.15465954]
 [0.8017439  0.19825609]
 [0.34065586 0.65934414]]

sklearn probabilities:  [[4.32596905e-04 9.99567403e-01]
 [9.99783238e-01 2.16761787e-04]
 [9.99783238e-01 2.16761787e-04]
 [4.32596905e-04 9.99567403e-01]
 [9.99783238e-01 2.16761787e-04]
 [9.99783238e-01 2.16761787e-04]
 [9.99783238e-01 2.16761787e-04]
 [9.99783238e-01 2.16761787e-04]
 [4.32596905e-04 9.99567403e-01]
 [9.99783238e-01 2.16761787e-04]
 [9.99783238e-01 2.16761787e-04]
 [4.32596905e-04 9.99567403e-01]]
```
As can be seen above, although both models have the same class predictions the sklearn model is clearly more confident. I tried to make the xgboost params match the sklearn default params as closely as possible (and turning off the lambda regularization in xgb) and I still get this big mis match. Any help would be appreciated, thanks.

**Comment by humzaiqbal** [comment]
[features_labels.zip](https://github.com/dmlc/xgboost/files/3331458/features_labels.zip)
The `features.npy` and `labels.npy` files are attached in the zip file

**Comment by hcho3** [comment]
@humzaiqbal GradientBoostingClassifier and XGBoost are different algorithms, so there is no good reason to expect them to behave identically. In particular,
* Scikit-learn uses first-order gradient, whereas XGBoost uses second-order.
* Scikit-learn uses Gini or Information Gain as split criteria, whereas XGBoost uses the loss reduction as split criteria.

**Comment by hcho3** [comment]
@humzaiqbal You can see the algorithmic details of XGBoost in my master's thesis: https://drive.google.com/file/d/0B0c0MbnP6Nn-eUNRRkVOOGpkbFk/view?usp=sharing

**Comment by trivialfis** [comment]
@hcho3 @humzaiqbal Actually I ran some experiments out of curiosity.  I just generated some random normal data, when the data has only 12 rows, the difference is similar to above listed result.  But if you try to increase the number of rows to 120, the difference will become much smaller.
